tests/unit/LTI_OIDC_Login_Test.php

```diff
@@ -119,4 +120,70 @@ public function testDoOidcLoginRedirect()
         $this->assertEquals('password123', $query['login_hint']);
         $this->assertEquals('my message hint', $query['lti_message_hint']);
     }
+
+    /**
+     * @dataProvider clientIdProvider
+     * @param array $request The request array to use for the test
+     */
+    public function testClientIdIsSetOnOidcLogin(array $request)
+    {
+        /** @var InMemoryCache|\PHPUnit_Framework_MockObject_MockObject */
+        $cache = $this->getMockBuilder(InMemoryCache::class)
+            ->setMethods(['cache_nonce'])
+            ->getMock();
+
+        $cache->expects($this->once())->method('cache_nonce')->with(
+            $this->stringStartsWith('nonce-')
+        );
+
+        /** @var Cookie|\PHPUnit_Framework_MockObject_MockObject $cookie */
+        $cookie = $this->getMockBuilder(Cookie::class)
+            ->setMethods(['set_cookie'])
+            ->getMock();
+
+        $cookie->expects($this->once())->method('set_cookie')->with(
+            $this->stringStartsWith('lti1p3_state-'),
+            $this->stringStartsWith('state-')
+        )->willReturn($cookie);
+
+        $redirect = LTI_OIDC_Login::newInstance(
+            new DummyDatabase(),
+            $cache,
+            $cookie
+        )->do_oidc_login_redirect(
+            $this->launchUrl,
+            $request
+        );
+
+        $expectedId = isset($request['aud']) ? $request['aud'] : $request['client_id'];
+
+        $this->assertInstanceOf(Redirect::class, $redirect);
+
+        $redirectUrl = parse_url($redirect->get_redirect_url());
+
+        parse_str($redirectUrl['query'], $query);
+        $this->assertEquals($expectedId, $query['client_id']);
+    }
+
+    public function clientIdProvider()
+    {
+        return [
+            'clientId as client_id' => [
+                [
+                    'iss' => 'aaa',
+                    'login_hint' => 'password123',
+                    'lti_message_hint' => 'my message hint',
+                    'client_id' => '12345'
+                ]
+            ],
+            'clientId as aud' => [
```

Comment:
Add one more condition here with both `client_id` and `aud` with conflicting values and confirm that it uses `client_id`